rrs/tools/upgrade_history_internal.py

```diff
@@ -0,0 +1,617 @@
+        prev_pv = latest_upgrade.version
+        prev_srcrev = latest_upgrade.srcrev
+    else:
+        prev_pv = None
+        prev_srcrev = ''
+
+    if prev_pv is None:
+        logger.debug("%s: Initial upgrade ( -> %s)." % (pn, pv))
+        _save_upgrade(recipesymbol, layerbranch, pv, srcrev, license, ct, title, info, filepath, logger)
+    else:
+        from common import get_recipe_pv_without_srcpv
+
+        (ppv, _, _) = get_recipe_pv_without_srcpv(prev_pv,
+                get_pv_type(prev_pv))
+        (npv, _, _) = get_recipe_pv_without_srcpv(pv,
+                get_pv_type(pv))
+
+        try:
+            vercmp_result = 0
+            if not (npv == 'git' or ppv == 'git'):
+                vercmp_result = vercmp_string(ppv, npv)
+
+            if npv == 'git':
+                logger.debug("%s: Avoiding upgrade to unversioned git." % pn)
+            elif ppv == 'git' or vercmp_result != 0 or srcrev != prev_srcrev or latest_upgrade.upgrade_type == 'R':
+                if initial is True:
+                    logger.debug("%s: Update initial upgrade ( -> %s)." % \
+                            (pn, pv)) 
+                    latest_upgrade.filepath = filepath
+                    latest_upgrade.version = pv
+                    latest_upgrade.save()
+                else:
+                    # Check if the "new" version is already in the database
+                    same_pv_upgrade = all_rupgrades.filter(version=pv).order_by('-commit_date', '-id').first()
+                    if same_pv_upgrade and \
+                            not all_rupgrades.filter(prev_version=pv, commit_date__gt=same_pv_upgrade.commit_date).exists() \
+                            and \
+                            not all_rupgrades.filter(upgrade_type__in=['R', 'N'], commit_date__gt=same_pv_upgrade.commit_date).exists():
+                        # The "previous" recipe is still present, we won't call this an upgrade
+                        logger.debug('%s: new version %s already exists' % (pn, pv))
+                        return
+                    upgrade_type = 'U'
+                    if vercmp_result == 1:
+                        upgrade_type = 'D'
+                    op = {'U': 'upgrade', 'D': 'downgrade'}[upgrade_type]
+                    logger.debug("%s: detected %s (%s -> %s)" \
+                            " in ct %s." % (pn, op, prev_pv, pv, ct))
+                    _save_upgrade(recipesymbol, layerbranch, pv, srcrev, license, ct, title, info, filepath, logger, upgrade_type=upgrade_type, orig_filepath=orig_filepath, prev_version=prev_pv)
+        except KeyboardInterrupt:
+            raise
+        except Exception as e:
+            logger.error("%s: fail to detect upgrade (%s -> %s)" \
+                            " in ct %s: %s" % (pn, prev_pv, pv, ct, str(e)))
+
+
+"""
+    Returns a list containing the fullpaths to the recipes from a commit.
+"""
+def _get_recipes_filenames(ct, repo, repodir, layersubdir_start, logger):
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this function is 20. By comparison, 98% of the functions in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the function has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this function or break it into multiple functions. For example, consider extracting the code block on lines 352-357 into a separate function.